base/mpfr.jl

```diff
@@ -280,11 +281,52 @@ BigFloat(x::Union{UInt8,UInt16,UInt32}, r::MPFRRoundingMode=ROUNDING_MODE[]; pre
 BigFloat(x::Union{Float16,Float32}, r::MPFRRoundingMode=ROUNDING_MODE[]; precision::Integer=DEFAULT_PRECISION[]) =
     BigFloat(Float64(x), r; precision=precision)
 
+function rational_to_bigfloat(
+    x::Rational{S},
+    ::Type{T},
+    requested_precision::Integer,
+    to_float::C,
+    romo::RoundingMode,
+) where {C<:Union{Type,Function},S<:Integer,T<:Integer}
+    s = Int8(sign(x))
+    a = abs(x)
+
+    num = numerator(a)
+    den = denominator(a)
+
+    # Handle special cases
+    num_is_zero = iszero(num)
+    den_is_zero = iszero(den)
+    if den_is_zero
+        num_is_zero && return to_float(NaN)
+        return to_float(s * Inf)
+    end
+    num_is_zero && return to_float(false)
+
+    components = rational_to_float_components(
+      num,
+      den,
+      requested_precision,
+      T,
+      RationalToFloatRoundingMode.R(romo, s),
+    )
+    mantissa = to_float(components.mantissa)
+
+    # TODO: faster?
+    ldexp(s * mantissa, components.exponent)
+end
+
 function BigFloat(x::Rational, r::MPFRRoundingMode=ROUNDING_MODE[]; precision::Integer=DEFAULT_PRECISION[])
     setprecision(BigFloat, precision) do
-        setrounding_raw(BigFloat, r) do
-            BigFloat(numerator(x))::BigFloat / BigFloat(denominator(x))::BigFloat
-        end
+        rational_to_bigfloat(
+            x,
+            BigInt,
+            precision,
+            let prec = precision
+                y -> BigFloat(y, r, precision = prec)
+            end,
```

Comment:
For clarification, you're asking for something like `rational_to_bigfloat(x, BigInt, precision, setprecision(...`?

Reply:
Actually I think that `setprecision` is only needed around `ldexp`, so I'll just put it around `ldexp`.

Reply:
Although I guess we actually don't need `ldexp` at all, it would be better to convert the `BigInt` "mantissa" to `BigFloat` and then multiply with a power of two. Or something like that anyway.

Reply:
Though then we still need `setprecision` for multiplying with a power of two.

Reply:
I'll just do ccall's directly, since we're in mpfr.jl anyway.

Reply:
Inline `ccall` is generally not a good idea - either wrap a as-of-yet unwrapped interface and call the julia function, or use the existing functionality to achieve the same goal. This makes it easier for others to read & follow the code, without knowing that `BigFloat` is using MPFR under the hood (which is an implementation detail).